File: backend/src/main/scala/merchant/api/UpdateMenuItemStockApi.scala
package merchant.api

import domain.shared.given

import cats.effect.IO
import domain.merchant.UpdateMenuItemStockRequest
import domain.shared.{DeliveryAppState, MenuItemId, StoreId, UserRole}
import merchant.app.*
import org.http4s.HttpRoutes
import org.http4s.circe.CirceEntityCodec.*
import org.http4s.dsl.io.*
import shared.api.routing.*
import shared.app.*

val updateMenuItemStockApi: FixedMethodApi2[StoreId, MenuItemId, DeliveryAppState] =
  jsonPostApi2[StoreId, MenuItemId, UpdateMenuItemStockRequest, DeliveryAppState](
    List(routeSegment("api"), routeSegment("delivery"), routeSegment("stores")),
    List(routeSegment("menu")),
    List(routeSegment("stock")),
  )

File: frontend/src/apis/merchant/UpdateMenuItemStockApi.ts
import type { DeliveryAppState, MenuItemId, StoreId, UpdateMenuItemStockRequest } from '@/objects/core/SharedObjects'
import { postNormalizedDeliveryState } from '@/system/api/DeliveryStateClient'
import { buildApiPath2, defineJsonPostApi2, routeSegment } from '@/system/api/TypedApiDefinitions'

export const updateMenuItemStockApiDefinition = defineJsonPostApi2<StoreId, MenuItemId, UpdateMenuItemStockRequest, DeliveryAppState>(
    [routeSegment('api'), routeSegment('delivery'), routeSegment('stores')],
    [routeSegment('menu')],
    [routeSegment('stock')],
  )

export function updateMenuItemStock(
  storeId: StoreId,
  menuItemId: MenuItemId,
  payload: UpdateMenuItemStockRequest,
) {
  return postNormalizedDeliveryState(
    buildApiPath2(updateMenuItemStockApiDefinition, storeId, menuItemId),
    payload,
  )
}
